### DISCUSSION TITLE
multi select that updates a graph based on (de)selections

### DISCUSSION THREAD

**Discussion by metaperl** [zauberzeug/nicegui#1752] [Q&A]
### Question

The function `example_3()` in the code below plots all 3 ticker symbols from the attached CSV file. I would like to extend `example_3` so that it plots any of the 3 symbols which are currently selected in a [multi select pulldown](https://nicegui.io/documentation/select).

Thus as things are selected and/or de-selected, I would like for the graph to only plot the ones currently selected. Code and CSV file follow:

```python
import pandas as pd

from nicegui import ui


def plot(datums):
    chart = ui.chart({
        'chart': {
            'type': 'spline',
        },
        'legend': {
            'enabled': True,
            'layout': 'vertical',
            'align': 'right',
            'verticalAlign': 'middle',
        },
        'navigator': {
            'enabled': True
        },
        'scrollbar': {
            'enabled': False
        },
        'series': datums
    }, type='stockChart', extras=['stock'])


def toy_example():
    data1 = {'name': 'alice', 'data': [31, 148, 33, 68, 75, 43]}
    data2 = {'name': 'bob', 'data': [3, 28, 7, 5, 8, 3]}
    data3 = {'name': 'john', 'data': [1, 4, 0, 8, 9, 0]}
    plot([data1, data2, data3])


def ticker_and_close(ticker, df):
    """Return a dictionary suitable for plotting.

    Looks through dataframe and returns a ticker and it's close data in a format like so:
    {'name': 'stock_symbol', 'data': [31, 148, 33, 68, 75, 43]}
    """

    _ = df[df['Ticker'] == ticker]
    data = {'name': ticker, 'data': _['Close'].tolist()}
    return data


def example_2():
    """Plot a single ticker in the CSV file."""

    source = pd.read_csv('ticker-data.csv')
    source['Date'] = pd.to_datetime(source['Date'])
    all_tickers = source.Ticker.unique()

    data = ticker_and_close('PTON', source)
    plot([data])

def example_3():
    """Plot all 3 tickers in CSV file."""

    source = pd.read_csv('ticker-data.csv')
    source['Date'] = pd.to_datetime(source['Date'])
    all_tickers = source.Ticker.unique()

    datums = [ticker_and_close(x, source) for x in "PTON TDOC ZM".split()]
    plot(datums)

# toy_example()
# example_2()
# example_3()

ui.run()
```
[ticker-data.csv](https://github.com/zauberzeug/nicegui/files/12822479/ticker-data.csv)

**Comment by falkoschindler** [comment]
Sorry, but your post doesn't include any question, only a request for writing code for you. Please try to narrow it down to the specific problem you're facing and tell us what you already tried (including code).

Besides that, your code example is pretty large. If you struggle with example 3, why including the others? Please try to reduce it to the bare minimum. And if at all possible, avoid the use of additional data files. This allows us and other contributors to help much more efficiently. Thanks! 🙂

**Reply by metaperl** [reply]
> Sorry, but your post doesn't include any question, only a request for writing code for you. Please try to narrow it down to the specific problem you're facing and tell us what you already tried (including code).

Yes, forcing me to try got me 98% of the way.

> If you struggle with example 3, why including the others? 

`example_3()` works :)

> Besides that, your code example is pretty large. If you struggle with example 3, why including the others?

I'm not struggling with 3. It works fine. But the goal was to make `example_3` render a subset of the tickers instead of all via a `select`.

> Please try to reduce it to the bare minimum. And if at all possible, avoid the use of additional data files. This allows us and other contributors to help much more efficiently.

This is a reasonable request. However, in the current case. I think one remaining problem will be easily identified just by glancing at the working code which is invoked via `example_4`. My code works just fine. But I have one problem: on each `on_change` event, it renders a new chart instead of overwriting the existing chart. How can I change this code so that it re-renders the data on the same chart instead of creating a new chart on each `on_change` event?

```python
import pandas as pd

from nicegui import ui

chart = None


def plot(datums):
    global chart

    chart = ui.chart({
        'chart': {
            'type': 'spline',
        },
        'legend': {
            'enabled': True,
            'layout': 'vertical',
            'align': 'right',
            'verticalAlign': 'middle',
        },
        'navigator': {
            'enabled': True
        },
        'scrollbar': {
            'enabled': False
        },
        'series': datums
    }, type='stockChart', extras=['stock'])


def ticker_and_close(ticker, df):
    """Return a dictionary suitable for plotting.

    Looks through dataframe and returns a ticker and it's close data in a format like so:
    {'name': 'stock_symbol', 'data': [31, 148, 33, 68, 75, 43]}
    """

    _ = df[df['Ticker'] == ticker]
    data = {'name': ticker, 'data': _['Close'].tolist()}
    return data

def plot_selected(selected):
    source = pd.read_csv('ticker-data.csv')
    source['Date'] = pd.to_datetime(source['Date'])
    datums = [ticker_and_close(x, source) for x in selected]
    plot(datums)


def example_4():
    source = pd.read_csv('ticker-data.csv')
    source['Date'] = pd.to_datetime(source['Date'])
    all_tickers = source.Ticker.unique()
    all_tickers_list = all_tickers.tolist()
    print(f"{all_tickers_list=}")

    ui.select(all_tickers_list, multiple=True, value=all_tickers_list,
              label='with chips', on_change=lambda e: plot_selected(e.value)
              ).classes('w-64').props('use-chips')

    plot_selected(all_tickers_list)

example_4()

ui.run()
```

**Reply by rodja** [reply]
You either need to update the chart with the new data instead of creating a new `ui.chart` in your `plot` function, or remove the old ui.chart element. The latter may be done with an [refreshable](https://nicegui.io/documentation/refreshable).

**Reply by metaperl** [reply]
>  update the chart with the new data

How would you update the chart with new data? In looking through the [chart docs](https://nicegui.io/documentation/chart) I dont see any mention of how to do this.

**Reply by metaperl** [reply]
Also as feedback I would like to say that `all_tickers_list=` is printed twice when the program is invoked.. when using a logging library, I detected that both `__main__` and `__mp_main__` invoke that line with the program implemented as follows;

```
2023-10-10 08:38:33.075 | DEBUG    | __main__:example_4:52 - all_tickers_list=['PTON', 'TDOC', 'ZM']
2023-10-10 08:38:34.864 | DEBUG    | __mp_main__:example_4:52 - all_tickers_list=['PTON', 'TDOC', 'ZM']
```

```python
import pandas as pd

from nicegui import ui

from loguru import logger

def plot(datums):
    ui.chart({
        'chart': {
            'type': 'spline',
        },
        'legend': {
            'enabled': True,
            'layout': 'vertical',
            'align': 'right',
            'verticalAlign': 'middle',
        },
        'navigator': {
            'enabled': True
        },
        'scrollbar': {
            'enabled': False
        },
        'series': datums
    }, type='stockChart', extras=['stock'])


def ticker_and_close(ticker, df):
    """Return a dictionary suitable for plotting.

    Looks through dataframe and returns a ticker and it's close data in a format like so:
    {'name': 'stock_symbol', 'data': [31, 148, 33, 68, 75, 43]}
    """

    _ = df[df['Ticker'] == ticker]
    data = {'name': ticker, 'data': _['Close'].tolist()}
    return data


def plot_selected(selected):
    source = pd.read_csv('ticker-data.csv')
    source['Date'] = pd.to_datetime(source['Date'])
    datums = [ticker_and_close(x, source) for x in selected]
    plot(datums)


def example_4():
    source = pd.read_csv('ticker-data.csv')
    source['Date'] = pd.to_datetime(source['Date'])
    all_tickers = source.Ticker.unique()
    all_tickers_list = all_tickers.tolist()
    logger.debug(f"{all_tickers_list=}")

    ui.select(all_tickers_list, multiple=True, value=all_tickers_list,
              label='with chips', on_change=lambda e: plot_selected(e.value)
              ).classes('w-64').props('use-chips')

    # plot_selected(all_tickers_list)

example_4()

ui.run()
```

**Reply by falkoschindler** [reply]
https://github.com/zauberzeug/nicegui/wiki/FAQs#why-is-my-code-executed-twice

To update a chart: Keep a reference to the options dictionary, modify it as you wish and call `chart.update()` once you're done.

**Reply by metaperl** [reply]
> To update a chart: Keep a reference to the options dictionary, modify it as you wish and call chart.update() once you're done.


1. Your answer about updating a chart sounds like another FAQ or perhaps something that should be in the chart documentation. dont you think?
2. You mean there is no way to access the options from what is returned back from calling `ui.chart(...)` ? I cant simply do `chart.options.series = new_series` ?
3. is also the case that the suggestion remove the old ui.chart element  with an [refreshable](https://nicegui.io/documentation/refreshable) was not a workable solution?

I dont think I'm going to be the last person to have the problems I am having. I am supplying a toy example below. I encourage you to put this in your docs somewhere after updating the `chart_update` function to exemplify what you mean by `To update a chart: Keep a reference to the options dictionary, modify it as you wish and call chart.update() once you're done.`

```python
from nicegui import ui

# initial data
data1 = {'name': 'alice', 'data': [31, 148, 33, 68, 75, 43]}

# data to update chart options with to exemplify updating a chart.
data2 = {'name': 'bob', 'data': [3, 28, 7, 5, 8, 3]}
data3 = {'name': 'john', 'data': [1, 4, 0, 8, 9, 0]}

chart = None
chart_options = {
    'chart': {
        'type': 'spline',
    },
    'legend': {
        'enabled': True,
        'layout': 'vertical',
        'align': 'right',
        'verticalAlign': 'middle',
    },
    'navigator': {
        'enabled': True
    },
    'scrollbar': {
        'enabled': False
    },
    'series': [data1]
}


def plot():
    global chart, chart_options
    chart = ui.chart(chart_options, type='stockChart', extras=['stock'])


def update_chart():
    global chart_options


plot()
ui.run(reload=False)
```

**Reply by falkoschindler** [reply]
@metaperl I was mistaken: There _is_ an `options` property you can use to update the chart.

As already written in the [documentation](https://nicegui.io/documentation/chart):
> Updates can be pushed to the chart by changing the `options` property. After data has changed, call the `update` method to refresh the chart.

And there is a code example demonstrating updates like this:
```py
def update():
    chart.options['series'][0]['data'][0] = random()
    chart.update()
```

**Reply by metaperl** [reply]
I have a toy example working now and will move on to the real code in a bit. Working toy example follows:

```python
from nicegui import ui

# initial data
data1 = {'name': 'alice', 'data': [31, 148, 33, 68, 75, 43]}

# data to update chart options with to exemplify updating a chart.
data2 = {'name': 'bob', 'data': [3, 28, 7, 5, 8, 3]}
data3 = {'name': 'john', 'data': [1, 4, 0, 8, 9, 0]}

chart = None
chart_options = {
    'chart': {
        'type': 'spline',
    },
    'legend': {
        'enabled': True,
        'layout': 'vertical',
        'align': 'right',
        'verticalAlign': 'middle',
    },
    'navigator': {
        'enabled': True
    },
    'scrollbar': {
        'enabled': False
    },
    'series': [data1]
}


def plot():
    global chart, chart_options
    chart = ui.chart(chart_options, type='stockChart', extras=['stock'])


def update_chart():
    global chart

    chart.options['series'] = [data2, data3]
    chart.update() 


plot()
ui.button('Update', on_click=update_chart)
ui.run(reload=False)
```

**Reply by metaperl** [reply]
actually I detected an issue with the toy example that is now reported as a bug - https://github.com/zauberzeug/nicegui/issues/1789
